The intricate nature of tumor heterogeneity poses significant challenges in developing effective cancer therapies. Single-cell RNA sequencing (scRNA-seq) has greatly advanced our understanding of the subtle variations within and between tumors, promoting the development of personalized treatments. Moreover, Spatial Transcriptomics (ST) has transformed sequencing methods by enabling precise localization of gene expression within intact tissue structures. Integrating ST into cancer research is crucial for unraveling the complexities of tumor heterogeneity, as it facilitates the identification of malignant traits in individual cells or specific regions.
Currently, numerous tools have emerged that can predict drug sensitivity at the single-cell level, significantly advancing personalized treatment approaches. However, the training processes of these tools often overlook the spatial distribution and interaction of drug-sensitive and drug-resistant cells within and beyond the Tumor-Immune Interface. This oversight can result in scenarios where drug-sensitive cells exhibit low spatial coherence while drug-resistant cells show high spatial coherence, potentially leading to suboptimal drug efficacy. Therefore, incorporating spatial coherence considerations into these predictive models is essential for improving the effectiveness of personalized cancer therapies.
